### PR TITLE
interface: convert state to hooks, part 2

### PR DESCRIPTION
**Pull request by tylershuster**
(part 2 of https://github.com/urbit/urbit/pull/4171)

This PR replaces the current monolithic global state subscription and reduction with a modular, hook-based state management per module.

Currently, only the invite state has been replaced for discussion and refinement of the pattern. Relevant files for examining this pattern:
```
~/logic/state/invite.ts
~/logic/lib/useApi.ts
~/logic/subscription/invite.ts
```

As a part of this process, this also introduces a new folder at `pkg/npm` which contains NPM packages for publication, currently these:
```
pkg/npm/http-api: A typescript library for interacting with eyre over http — currently lives on my personal repo
pkg/npm/api: Typescript types for generic (non-Landscape) data — new, extracted from pkg/interface/src/types
```
It requires these via relative import, precluding the need for publishing these packages during development. 

@matildepark We should discuss methods for automatically deploying these packages on NPM when they have been updated.

@liam-fitzgerald & @tacryt-socryp code review would be much appreciated before I convert the other apps.

@ixv As you can see, settings remain local, but the files listed above should give you an idea of the pattern I hope to use.

This has some holes in it, especially around packaging. You may have to run `npm run build` in the `npm/http-api` folder; let me know. There are some other obvious problems — don't hesitate to point them out, but don't feel like you need to spend your time pointing out unnecessary imports and the like.

### CODE REVIEW

File: pkg/interface/src/logic/lib/useApi.ts
Comment (tylershuster) [inline pkg/interface/src/logic/lib/useApi.ts:11]: Programmatic channel creation when

File: pkg/interface/src/logic/subscription/invite.ts
Comment (tylershuster) [inline pkg/interface/src/logic/subscription/invite.ts:68]: This feels like there's a slightly more elegant way to do this, but the basic idea is that `old state + update operation = new state`, with some javascript on top to make it work

File: pkg/interface/src/views/App.js
Comment (tylershuster) [inline pkg/interface/src/views/App.js:89]: Just gets the ball rolling asap

File: pkg/interface/src/views/apps/chat/components/content/text.js
Comment (tylershuster) [inline pkg/interface/src/views/apps/chat/components/content/text.js:32]: This is accidental

File: pkg/npm/api/index.d.ts
Comment (tylershuster) [inline pkg/npm/api/index.d.ts:32]: Really?

File: pkg/npm/http-api/LICENSE
Comment (tylershuster) [inline pkg/npm/http-api/LICENSE:3]: Probably not

File: pkg/npm/http-api/example/browser.js
Comment (tylershuster) [inline pkg/npm/http-api/example/browser.js:1]: These should probably not be checked in, but *should* be published to NPM. @matildepark

File: pkg/interface/src/logic/state/invite.tsx
Comment (tylershuster) [inline pkg/interface/src/logic/state/invite.tsx:27]: I'm not sure these need to be here. They could be replaced by `useApi` calls and pokes at the relevant location in the component which needs to do it.

File: pkg/npm/api/package.json
Comment (liam-fitzgerald) [inline pkg/npm/api/package.json:10]: MIT?

File: pkg/interface/src/logic/lib/useApi.ts
Comment (liam-fitzgerald) [reply pkg/interface/src/logic/lib/useApi.ts:11]: A subscription also opens the channel?

File: pkg/interface/src/logic/lib/useApi.ts
Comment (liam-fitzgerald) [inline pkg/interface/src/logic/lib/useApi.ts:6]: This doesn't need to be async

File: pkg/interface/src/logic/subscription/invite.ts
Comment (liam-fitzgerald) [reply pkg/interface/src/logic/subscription/invite.ts:68]: Why the `bind`?

File: pkg/npm/http-api/src/index.ts
Comment (liam-fitzgerald) [inline pkg/npm/http-api/src/index.ts:100]: double return??

File: pkg/npm/http-api/src/index.ts
Comment (liam-fitzgerald) [inline pkg/npm/http-api/src/index.ts:183]: You should probably rethrow here

File: pkg/npm/http-api/src/index.ts
Comment (liam-fitzgerald) [inline pkg/npm/http-api/src/index.ts:260]: Not blocking, but this should surface reconnect functionality a la channel.js

File: pkg/npm/http-api/src/index.ts
Comment (liam-fitzgerald) [inline pkg/npm/http-api/src/index.ts:351]: why bother with both callback style and promises?

File: pkg/npm/http-api/src/index.ts
Comment (liam-fitzgerald) [inline pkg/npm/http-api/src/index.ts:360]: Should rethrow

File: pkg/npm/http-api/yarn.lock
Comment (liam-fitzgerald) [inline pkg/npm/http-api/yarn.lock:2]: lockfiles shouldn't be committed for libraries

**Review by liam-fitzgerald — CHANGES_REQUESTED** [review]
Very excited to have this in. Have some questions about the design of the libraries. What does the build process look like for these dependent packages? Also, you should probably consider error handling.

File: pkg/npm/api/package.json
Comment (tylershuster) [reply pkg/npm/api/package.json:10]: Leftover boilerplate

File: pkg/interface/src/logic/subscription/invite.ts
Comment (tylershuster) [reply pkg/interface/src/logic/subscription/invite.ts:68]: All of the reducers are operating on the same incoming data but the state changes as it passes through

File: pkg/npm/http-api/src/index.ts
Comment (tylershuster) [reply pkg/npm/http-api/src/index.ts:100]: I had put in the proxy technique but it's ugly and of questionable usefulness. I left the old code in in case I found a reason to go back.

File: pkg/interface/src/logic/lib/useApi.ts
Comment (liam-fitzgerald) [reply pkg/interface/src/logic/lib/useApi.ts:6]: for example: 

```typescript
import Urbit from '@urbit/http-api';

import { memoize } from "lodash";
import inviteSubscription from '../subscription/invite';

const useApi = memoize(() => {
  const channel = new Urbit(window.location.origin, '');
  (async () => {
    channel.verbose = true;
    channel.ship = window.ship;
    channel.subscribe(inviteSubscription(channel));
    await channel.poke({ app: 'hood', mark: 'helm-hi', json: 'Loading Landscape' });
    await channel.eventSource(); 
  })();

  return channel;

});

export default useApi;

```

**Comment by tylershuster** [comment]
@liam-fitzgerald your work in https://github.com/urbit/urbit/pull/4285/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx

and

@tacryt-socryp your work in https://github.com/urbit/urbit/pull/4264/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx

will both have to be adapted into the new pattern, presuming this lands *after* your changes. The changes in api/metadata::preview make me a little leery but I imagine it's nothing that can't be overcome. I'm mostly leaving this comment here for myself to reference after those PRs get merged

**Comment by tylershuster** [comment]
@matildepark are we just waiting until `release/next-userspace` gets merged now?

**Comment by matildepark** [comment]
Alas yes... if there’s stuff that can be merged without breaking much it can go in by itself. For now too many broad changes at once for a huge pr like this.

**Comment by tylershuster** [comment]
No, that's why I opened https://github.com/urbit/urbit/pull/4338

**Comment by tylershuster** [comment]
@liam-fitzgerald I'll address these comments in my other PR

**Pull request closed without merging**
